src/coap_openssl.c

```diff
@@ -492,138 +510,724 @@ int coap_dtls_context_set_psk(coap_context_t *ctx,
+      int pkey_type = map_key_type(setup_data->key.asn1.private_key_type);
+      if (!(SSL_use_PrivateKey_ASN1(pkey_type, ssl,
+            setup_data->key.asn1.private_key, setup_data->key.asn1.private_key_len))) {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure Client Private Key\n",
+                 "ASN1");
         return 0;
       }
     }
     else {
-      coap_log(LOG_ERR, "*** coap_dtls_context_set_pki: TLS: No Server Private Key defined\n");
+      coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: No Client Private Key defined\n");
+      return 0;
+    }
+    if (setup_data->key.asn1.ca_cert && setup_data->key.asn1.ca_cert_len > 0) {
+      /* Need to use a temp variable as it gets incremented*/
+      const uint8_t *p = setup_data->key.asn1.ca_cert;
+      X509* x509 = d2i_X509(NULL, &p, setup_data->key.asn1.ca_cert_len);
+      X509_STORE *st;
+      long e;
+      SSL_CTX *ctx = SSL_get_SSL_CTX(ssl);
+
+      /* We do not add CA for client to send to client as we are client! */
+
+      /* Add CA to the trusted root CA store */
+      st = SSL_CTX_get_cert_store(ctx);
+      /* Flush out existing errors */
+      while ((e = ERR_get_error()) != 0) { }
+      if (!X509_STORE_add_cert(st, x509)) {
+        while ((e = ERR_get_error()) != 0) {
+          int r = ERR_GET_REASON(e);
+          if (r != X509_R_CERT_ALREADY_IN_HASH_TABLE) {
+            coap_log(LOG_WARNING, "*** setup_pki: (D)TLS: %s at %s:%s\n",
+                     ERR_reason_error_string(e),
+                     ERR_lib_error_string(e),
+                     ERR_func_error_string(e));
+          }
+        }
+      }
+      X509_free(x509);
+    }
+    break;
+  default:
+    coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: Unknown key type %d\n",
+             setup_data->key_type);
+    return 0;
+  }
+  return 1;
+}
+
+static char*
+get_common_name_from_cert(X509* x509) {
+  if (x509) {
+    char *cn;
+    int n;
+    STACK_OF(GENERAL_NAME) *san_list;
+    char buffer[256];
+
+    san_list = X509_get_ext_d2i(x509, NID_subject_alt_name, NULL, NULL);
+    if (san_list) {
+      int san_count = sk_GENERAL_NAME_num(san_list);
+
+      for (n = 0; n < san_count; n++) {
+        const GENERAL_NAME * name = sk_GENERAL_NAME_value (san_list, n);
+
+        if (name->type == GEN_DNS) {
+          const char *dns_name = (const char *)ASN1_STRING_get0_data(name->d.dNSName);
+
+          /* Make sure that there is not an embedded NUL in the dns_name */
+          if (ASN1_STRING_length(name->d.dNSName) != (int)strlen (dns_name))
+            continue;
+          cn = OPENSSL_strdup(dns_name);
+          sk_GENERAL_NAME_pop_free(san_list, GENERAL_NAME_free);
+          return cn;
+        }
+      }
+      sk_GENERAL_NAME_pop_free(san_list, GENERAL_NAME_free);
+    }
+    /* Otherwise look for the CN= field */
+    X509_NAME_oneline(X509_get_subject_name(x509), buffer, sizeof(buffer));
+    cn = strcasestr(buffer, "CN=");
```

Comment:
`strcasestr()` is a nonstandard extension. The idea is to keep these out of the core library.

Reply:
Fixed by using local equivalent